### ISSUE TITLE
FR:  Some ideas to extend the functionality of Newsdownloader

### ISSUE THREAD

**Issue by roygbyte** [koreader/koreader#9225]
I've been working on a rebuild of Newsdownloader. After my recent failed refactoring attempt, I decided that it would be best for my sanity (and for the plugin) if it were started over from scratch. So far it's going well, and you can see the current progress on [my repo](https://github.com/roygbyte/gazette.koplugin).

I'm trying to design the code such that it can accommodate new functionality without over complicating the bread-and-butter of the plugin. Obviously, we want it to download RSS feeds and convert them to Epubs. :) Less obviously, we might want to have a bit of control over _how_ that happens. Here are some ideas:

- Option to set per-feed download locations.
- Ability to group feeds together, and have the feeds' new posts downloaded into a single Epub (similar to the "volumization" feature request).
- Ability to do per-feed downloading (or in the case above, per-group downloading)
- Ability to create an "infinite epub" (new posts are appended to an existing Epub).
- Visual indicator or interface to show which feeds have downloaded new posts.

I can't promise I will actually implement these features. I'm mostly putting them out there in this FR to gather feedback, and to remind myself to design the code so that these features could hypothetically be implemented. 

Let me know what you think!

**Comment by poire-z** [comment]
May be ping the people who gave feedback about your previous attempt, or that filled issues about newsdownloader in the past - so you can discuss what they wish / don't need.

I remember I suggested a few things in the past in various issues that I don't want to search for - but found this one: https://github.com/koreader/koreader/pull/8145#issuecomment-927140425. (How's hell? :)

**Comment by roygbyte** [comment]
Good idea... I'll chase down some of the old threads and see what people think. 

Hell is actually not so bad this time of the year! Now that I've seen through to the other side of some bizarre errors (the result of operator error, of course) and a new-to-to-me way of structuring code (avoiding dependency injection and using factories) it's actually going really well :) I expect when I tread into LuaSettings land I'll be having an awful time, though. So be sure to visit again soon.

**Comment by arooni** [comment]
Love to see this!  

Would love to get this working on a Kindle Paperwhite.  Would be very handy!  

I seem to recall that the error messages were a bit challenging to decipher when there were poorly formed RSS urls.  it would be nice if it could fail gracefully and not prevent other sources to fail.

**Comment by roygbyte** [comment]
@arooni that's a good idea. When a feed fails to download It would be helpful to know if:
- The feed website could be reached
- The feed website returned an `application/xml` response
- The feed was properly formatted (e.g.: a valid RSS feed)
- The items/entries in the feed exist
I am imagining a class (`FeedError` ?) that takes a request/response combination, deciphers them, and figures out what the heck went wrong!

**Comment by Frenzie** [comment]
Not sure if #3428 (e.g., start filename with a date like `2018-01-15T11:10:36Z`) is still relevant?

**Comment by roygbyte** [comment]
> Not sure if #3428 (e.g., start filename with a date like `2018-01-15T11:10:36Z`) is still relevant?

No idea. 

I'm intending to allow optional file name overriding with some user supplied formatting. Maybe this addresses the issue?

**Comment by Frenzie** [comment]
If the user supplied formatting is the parsed date, I think it would.

**Comment by roygbyte** [comment]
Does anyone have any big opinions about a "gazette" mode for newsdownloader? The idea being that you can join a number of feeds together into a single publication. Creating the publication would include the following steps:

- Create new publication
- Add subscriptions (e.g.: `ourworldindata.org/atom.xml`, `someotherfeed.com/rss.xml`)
- Set name of publication

Use cases:

- A person wants to amalgamate all their daily reads into a single volume, like a newspaper.
- A person wants to group subscriptions of a similar topic together, like a magazine.

**Comment by roygbyte** [comment]
Separate idea from above: setting unique download destinations for each feed.

Use cases:
- A person wants to put subscriptions for topic `x` stuff in their topic `x` folder.

**Comment by poire-z** [comment]
> Does anyone have any big opinions about a "gazette" mode for newsdownloader? The idea being that you can join a number of feeds together into a single publication.

No big opinion - but that sounds nice :)
(Mostly because if I wanted to read a few rss feeds, I'd rather have them all in a single book to not polllute my History.)

**Comment by roygbyte** [comment]
For this mythical Gazette mode to work, there would have to be a slight restructuring of the experience of adding a feed. Here's one take on that:

Step 1: setup the feed by creating a new "subscription". The "subscription" would handle all the configuration for the feed, like whether to download images, how to filter the webpage. The same feed could even have two separate subscriptions, in case a person wanted to use it for different gazette publications... which brings me to step 2.

Step 2: setup gazette mode by creating a new "publication". The "publication" can have one or more "subscriptions" added. This is the basic gazette mode feature.

**Comment by roygbyte** [comment]
Progress:

![Screenshot from 2022-09-14 19-45-26](https://user-images.githubusercontent.com/82218266/190278702-042ddb20-c319-40b7-9f03-fab9ca0b3d40.png)
![Screenshot from 2022-09-14 19-45-44](https://user-images.githubusercontent.com/82218266/190278701-992a6535-a62b-41d8-9b58-d22eeb0311cb.png)
![Screenshot from 2022-09-14 19-46-01](https://user-images.githubusercontent.com/82218266/190278687-09a9a421-c2ea-4488-af17-fc538d06efb8.png)
![Screenshot from 2022-09-14 19-52-27](https://user-images.githubusercontent.com/82218266/190278699-1596d344-7690-4b21-afa1-38bc001c90d2.png)
![Screenshot from 2022-09-14 19-52-39](https://user-images.githubusercontent.com/82218266/190278698-daea93bf-7d7a-4387-9d68-d71aefa9d186.png)

**Comment by roygbyte** [comment]
@Lukas1h Let me know how it goes... it's still pretty buggy ._.

**Comment by Lukas1h** [comment]
Installed the patches and added the plugin, but couldn't find it. 

> Edit: just realized I hadn't added some important files for the results view to the plugin. Sorry!

That's probably why. No problem I'll try again in a bit.

**Comment by Lukas1h** [comment]
Any idea how I would install the patches on Android? It looks like most of the `.lua` files are missing in the `koreader` folder.

**Comment by roygbyte** [comment]
I don't know anything about Android, I'm afraid! All the same: thank you for the reminder that people might like to use this, haha! There are some improvements I'd like to make and this is a small encouragement. :)

**Comment by roygbyte** [comment]
So I don't think the patches are required anymore. There were some PR that added the functionality.

**Comment by roygbyte** [comment]
When you're initially setting up the feed, I believe there's some guidance around what options are available for downloading the article or excerpt.

**Comment by sith-on-mars** [comment]
Thanks for the reply! I haven't be able to find the option to download full article. Could you guide me through it?

![image](https://user-images.githubusercontent.com/37974764/227053991-5ee84273-36dd-47b0-9ab6-d570d5cc14ac.png)

**Comment by roygbyte** [comment]
@sith-on-mars Sure! Looks like these are the three options:

``` lua
Feed.CONTENT_SOURCE = {
   SUMMARY = "summary",
   CONTENT = "content",
   WEBPAGE = "webpage",
}
```

Summary is the entry summary, as given by the feed. Content is the entry content, as given by the feed. Webpage is going to get the entire webpage and use that as the entry.

Thanks for trying this out! It's in such a chaotic state, I cannot remember what's working or what's broken. Good luck! Let me know how it goes :D

**Comment by sith-on-mars** [comment]
Thanks so much for the detailed guide! Switching to 'webpage' works for me to download full articles!

**Comment by sith-on-mars** [comment]
Hi roygbyte! I run into the problem with downloading the [RSS feed of London Review of Books](lrb.co.uk/feeds/rss). I was able to add the feed in my subscription, but it failed to synchronise any article. Basically, the synchronisation was stuck when downloading the first article. I tried both on 'content' and 'webpage', but to no avail. Interestingly, the stock RSS software also had the problem. Is there any workaround or could you look into this?

It's a wonderful plugin and I do find it quite useful!

**Comment by roygbyte** [comment]
Hi sith-on-mars, I'll look into it! Could be that that feed has strange formatting. I'm glad you like the plugin. I need to dig back into the code soon and "make it better" :)

**Comment by roygbyte** [comment]
@sith-on-mars I believe the issue is fixed by [trimming spaces from the title](https://github.com/koreader/koreader/pull/9637/commits/38e06849c9f574f06d2f1f8b96c089a2ebd3055e).

P.s.: I'm going to move development of this plugin back to my repo, because I don't think it will be merged anytime soon and that will make maintaining it easier. Will update this issue when I've done that.

**Comment by drcarademono** [comment]
Is this still under development? It's easily my most desired feature for KOReader.

**Comment by roygbyte** [comment]
drcardemono, not actively. I've lost the thread on this one. I'd like to come back around and refactor it and fix it up, but I'm not sure when I'll find the time.
